transformations/transformations/derived_types.py

```diff
@@ -28,148 +33,502 @@
+        dependencies_updated = self.expand_derived_args_caller(routine, successors_data)
+
+        # ...and invalidate cached properties if dependencies have changed...
+        if dependencies_updated and item:
+            item.clear_cached_property('imports')
 
-        # Apply argument transformation, caller first!
-        self.flatten_derived_args_caller(routine)
+        # ...before updating the routine's signature and replacing
+        # use of members in the body
         if role == 'kernel':
-            self.flatten_derived_args_routine(routine)
+            # Expand derived type arguments in kernel
+            orig_argnames = tuple(arg.lower() for arg in routine.argnames)
+            expansion_map = self.expand_derived_args_kernel(routine)
+            trafo_data = {
+                'orig_argnames': orig_argnames,
+                'expansion_map': expansion_map,
+            }
+            if item:
+                item.trafo_data[self._key] = trafo_data
 
-    @staticmethod
-    def _derived_type_arguments(routine):
+            if any('recursive' in prefix.lower() for prefix in routine.prefix or ()):
+                self.expand_derived_args_recursion(routine, trafo_data)
+
+
+    def expand_derived_args_caller(self, routine, successors_data):
         """
-        Find all derived-type arguments used in a given routine.
+        For all active :any:`CallStatement` nodes, apply the derived type argument
+        expansion on the caller side.
 
-        :return: A map of ``arg => [type_vars]``, where ``type_var``
-                 is a :class:`Variable` for each derived sub-variable
-                 defined in the original compound type.
+        The convention used is: ``derived%var => derived_var``.
+
+        Parameters
+        ----------
+        routine : :any:`Subroutine`
+            The routine in which to transform call statements
+        successors_data : :any:`CaseInsensitiveDict` of (str, dict)
+            Dictionary containing the expansion maps (key ``'expansion_map'``) and
+            original argnames (key ``'orig_argnames'``) of every child routine
+
+        Returns
+        -------
+        bool
+            Flag to indicate that dependencies have been changed (e.g. via new imports)
         """
-        # Get all variables used in the kernel that have parents
-        variables = FindVariables(unique=True).visit(routine.ir)
-        variables = [v for v in variables if hasattr(v, 'parent') and v.parent is not None]
-        candidates = defaultdict(list)
+        other_symbols = set()
+        call_mapper = {}
+        for call in FindNodes(CallStatement).visit(routine.body):
+            if not call.not_active:
+                call_name = str(call.name)
+                if call_name in successors_data:
+                    # Set the new call signature on the IR node
+                    arguments, kwarguments, others = self.expand_call_arguments(call, successors_data[call_name])
```

Comment:
Agreed, naming isn't particularly good. That's other symbols that come in via imports - need to double-check but maybe this can actually be removed...

Reply:
Yes, this wasn't necessary anymore